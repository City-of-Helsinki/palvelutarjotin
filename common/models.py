import uuid

from django.db import models
from django.utils.translation import ugettext_lazy as _


class TimestampedModel(models.Model):
    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    updated_at = models.DateTimeField(verbose_name=_("updated_at"), auto_now=True)

    class Meta:
        abstract = True


class UUIDPrimaryKeyModel(models.Model):
    id = models.UUIDField(
        verbose_name=_("UUID"), primary_key=True, default=uuid.uuid4, editable=False
    )

    class Meta:
        abstract = True
