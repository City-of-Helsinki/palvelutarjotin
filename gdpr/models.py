from django.db import models

from gdpr.consts import CLEARED_VALUE


class GDPRModel(models.Model):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        # raise a NotImplementedError
        # immediately on launch of an app
        # if the `gdpr_sensitive_data_fields` is not defined.
        self.gdpr_sensitive_data_fields

    class Meta:
        abstract = True

    @property
    def gdpr_sensitive_data_fields(self) -> list[str]:
        raise NotImplementedError(
            "The model that is extending abstract GDPRModel class "
            "must define a `gdpr_sensitive_data_fields` -field "
            "with names of the GDPR sensitive fields."
        )

    def clear_gdpr_sensitive_data_fields(self):
        """Clears all the GDPR data fields set by
        the static ``get_gdpr_sensitive_data_fields`` property
        and then calls ``save()`` for the model instance.

        If the field is blankable, set an empty string as an value.
        Else if the field is nullable, set None as a value.
        Else, set a ``CLEARED_VALUE`` constant as a value.

        NOTE: This method should be overridden if
        anything else is needed or wanted.
        """
        for field in self.gdpr_sensitive_data_fields:
            if self._meta.get_field(field).blank:
                setattr(self, field, "")
            elif self._meta.get_field(field).null:
                setattr(self, field, None)
            else:
                setattr(self, field, CLEARED_VALUE)
        self.save()
