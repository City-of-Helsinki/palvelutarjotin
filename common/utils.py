from datetime import datetime

import graphene
from django.conf import settings
from django.core.exceptions import PermissionDenied
from django.db import transaction
from django.utils import timezone
from graphene import Node
from graphql_relay import from_global_id

from palvelutarjotin import __version__
from palvelutarjotin.exceptions import (
    DataValidationError,
    IncorrectGlobalIdError,
    ObjectDoesNotExistError,
)
from palvelutarjotin.settings import REVISION

LanguageEnum = graphene.Enum(
    "Language", [(l[0].upper(), l[0]) for l in settings.LANGUAGES]
)

LINKED_EVENT_DATE_FORMAT = "%Y-%m-%d"


def format_linked_event_date(datetime_obj):
    return datetime_obj.strftime(LINKED_EVENT_DATE_FORMAT)


def update_object(obj, data):
    if not data:
        return
    for k, v in data.items():
        if v is None and not obj.__class__._meta.get_field(k).null:
            raise DataValidationError(f"{k} cannot be null.")
        setattr(obj, k, v)
    obj.save()


@transaction.atomic
def update_object_with_translations(model, model_data):
    translations_input = model_data.pop("translations", None)
    if translations_input:
        model.create_or_update_translations(translations_input)
    update_object(model, model_data)


def get_api_version():
    return " | ".join((__version__, REVISION.decode("utf-8")))


def get_node_id_from_global_id(global_id, node_name):
    name, id = from_global_id(global_id)
    if name != node_name:
        raise IncorrectGlobalIdError("Node type does not match")
    return id


def convert_to_localtime_tz(value):
    dt = datetime.combine(datetime.now().date(), value)
    if timezone.is_naive(value):
        # Auto add local timezone to naive time
        return timezone.make_aware(dt).timetz()
    else:
        return timezone.localtime(dt).timetz()


def get_obj_from_global_id(info, global_id, expected_obj_type):
    obj = Node.get_node_from_global_id(info, global_id)
    if not obj or type(obj) != expected_obj_type:
        raise ObjectDoesNotExistError(
            f"{expected_obj_type.__name__} matching query does not exist."
        )
    return obj


def get_editable_obj_from_global_id(info, global_id, expected_obj_type):
    obj = get_obj_from_global_id(info, global_id, expected_obj_type)
    try:
        if obj.is_editable_by_user(info.context.user):
            return obj
    except AttributeError:
        raise TypeError(
            f"{obj.__class__.__name__} model does not implement is_editable_by_user()."
        )
    raise PermissionDenied(
        f"User does not have permission to edit this {expected_obj_type.__name__}"
    )
