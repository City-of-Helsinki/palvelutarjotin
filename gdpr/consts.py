CLEARED_VALUE = "gdpr:cleared"
