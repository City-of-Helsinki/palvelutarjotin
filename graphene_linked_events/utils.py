# https://stackoverflow.com/questions/6578986/how-to-convert-json-data
# -into-a-python-object/15882054#15882054
import json
from collections import namedtuple
from django.conf import settings
from geopy import distance as geopy_distance
from geopy import Point
from typing import List

from graphene_linked_events.rest_client import LinkedEventsApiClient
from palvelutarjotin.exceptions import ApiBadRequestError, ObjectDoesNotExistError

api_client = LinkedEventsApiClient(config=settings.LINKED_EVENTS_API_CONFIG)


def format_response(response):
    """Remove @ sign from API response JSON keys."""

    def format_values(value):
        if isinstance(value, list):
            return [format_values(v) for v in value]
        elif isinstance(value, dict):
            return {
                key.replace("@", "internal_"): format_values(value)
                for key, value in value.items()
            }
        return value

    formatted = format_values(json.loads(response.text))
    return json.dumps(formatted)


def json_object_hook(d):
    return namedtuple("X", d.keys())(*d.values())


def json2obj(data):
    return json.loads(data, object_hook=json_object_hook)


def format_request(request):
    # TODO: Find better way to replace internal_id key
    return json.dumps(request).replace("internal_", "@")


def retrieve_linked_events_data(resource, resource_id, params=None, is_staff=False):
    response = api_client.retrieve(
        resource, resource_id, params=params, is_staff=is_staff
    )

    if response.status_code == 400:
        raise ApiBadRequestError("Could not establish a connection to the API.")

    if response.status_code == 404:
        raise ObjectDoesNotExistError("Could not find the event from the API.")

    # Raise any other errors
    response.raise_for_status()

    return json2obj(format_response(response))


def get_keyword_set_by_id(keyword_set_id):
    params = {"include": "keywords"}
    return retrieve_linked_events_data(
        "keyword_set",
        keyword_set_id,
        params=params,
    )


def bbox_for_coordinates(
    longitude: float, latitude: float, distance: float
) -> List[float]:
    """
    Calculate bounding box for the given coordinates so that the
    bounding box corner is at the given distance (in km).
    """
    sw_bearing = 225
    ne_bearing = 45
    d = geopy_distance.distance(kilometers=distance)

    starting_point = Point(latitude=latitude, longitude=longitude)
    sw_point = d.destination(starting_point, bearing=sw_bearing)
    ne_point = d.destination(starting_point, bearing=ne_bearing)

    return [
        sw_point.longitude,
        sw_point.latitude,
        ne_point.longitude,
        ne_point.latitude,
    ]
