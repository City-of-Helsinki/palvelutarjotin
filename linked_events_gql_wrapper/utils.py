# https://stackoverflow.com/questions/6578986/how-to-convert-json-data
# -into-a-python-object/15882054#15882054
import json
from collections import namedtuple


def format_response(response):
    # Some fields from api have @prefix that need to be converted
    return response.content.decode("utf-8").replace("@", "internal_")


def _json_object_hook(d):
    return namedtuple("X", d.keys())(*d.values())


def json2obj(data):
    return json.loads(data, object_hook=_json_object_hook)
