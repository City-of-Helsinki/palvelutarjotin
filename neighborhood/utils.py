import json
from collections import namedtuple

from django.conf import settings
from neighborhood.rest_client import NeighborhoodApiClient

api_client = NeighborhoodApiClient(config=settings.NEIGHBORHOOD_API_CONFIG)


def format_response(response):
    return response.text


def json_object_hook(d):
    return namedtuple("X", d.keys())(*d.values())


def json2obj(data):
    return json.loads(data, object_hook=json_object_hook)


def format_request(request):
    return json.dumps(request)
