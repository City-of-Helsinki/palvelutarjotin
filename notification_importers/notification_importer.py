import csv
import glob
import io
import os
import re
from collections import defaultdict
from logging import getLogger
from typing import DefaultDict, Dict, Mapping, Optional, Sequence, Tuple

import requests
from django.conf import settings
from django.core.exceptions import ObjectDoesNotExist
from django.db import transaction
from django_ilmoitin.dummy_context import dummy_context
from django_ilmoitin.models import NotificationTemplate, NotificationTemplateException
from django_ilmoitin.utils import render_notification_template
from parler.utils.context import switch_language
from requests import RequestException

# {
#     "<notification type>": {
#         "<language code>": {
#             "<field name>": "<field value>",
#             ...
#         },
#         ...
#     },
#     ...
# }
SourceData = Mapping[str, Mapping[str, Mapping[str, str]]]


logger = getLogger(__name__)


class NotificationImporterException(Exception):
    pass


class AbstractNotificationImporter:
    """
    An abstract base class for Notification Importer classes.
    Original source:
    https://github.com/City-of-Helsinki/kukkuu/blob/master/importers/notification_importer.py
    1. Some function and variable renaming done.
    2. Splitted code from original importer to create an abstract class
    """

    LANGUAGES = settings.PARLER_SUPPORTED_LANGUAGE_CODES
    FIELDS = ("subject", "body_text", "body_html")

    @transaction.atomic()
    def create_missing_and_update_existing_notifications(self) -> Tuple[int, int]:
        num_of_created = self.create_missing_notifications()
        num_of_updated = self.update_notifications(NotificationTemplate.objects.all())
        return num_of_created, num_of_updated

    @transaction.atomic()
    def create_missing_notifications(self) -> int:
        new_types = set(self.source_data.keys()) - set(
            NotificationTemplate.objects.values_list("type", flat=True)
        )
        for new_type in sorted(new_types):
            new_notification = NotificationTemplate(type=new_type)
            self._create_or_update_notification_using_source_data(new_notification)

        return len(new_types)

    @transaction.atomic()
    def update_notifications(
        self, notifications: Sequence[NotificationTemplate]
    ) -> int:
        num_of_updated = 0

        for notification in sorted(notifications, key=lambda n: n.type):
            if self.is_notification_in_sync(notification) is False:
                self._create_or_update_notification_using_source_data(notification)
                num_of_updated += 1

        return num_of_updated

    def is_notification_in_sync(
        self, notification: NotificationTemplate
    ) -> Optional[bool]:
        if notification.type not in self.source_data:
            return None

        for language in self.LANGUAGES:
            try:
                translation_obj = notification.translations.get(language_code=language)
            except ObjectDoesNotExist:
                translation_obj = None

            for field in self.FIELDS:
                current_value = (
                    self.clean_text(getattr(translation_obj, field))
                    if translation_obj
                    else ""
                )
                source_value = self._get_value_from_source(
                    notification.type, field, language
                )

                if current_value != source_value:
                    return False

        return True

    def _get_value_from_source(
        self, notification_type: str, field: str, language: str
    ) -> str:
        return (
            self.source_data.get(notification_type, {}).get(language, {}).get(field, "")
        )

    def _create_or_update_notification_using_source_data(
        self, notification: NotificationTemplate
    ):
        creating = not bool(notification.pk)

        for language in self.LANGUAGES:
            with switch_language(notification, language):
                for field in self.FIELDS:
                    setattr(
                        notification,
                        field,
                        self._get_value_from_source(notification.type, field, language),
                    )
                notification.save()

            # Test that the notification can be rendered without errors.
            # This create/update method is called inside a transaction so the
            # notification will not get actually saved in case of an error.
            try:
                render_notification_template(
                    notification,
                    context=dummy_context.get(notification.type),
                    language_code=language,
                )
            except NotificationTemplateException as e:
                raise NotificationImporterException(
                    f'Error rendering notification "{notification}": {e}'
                ) from e

        logger.info(
            f'{"Created" if creating else "Updated"} notification "{notification}"'
        )

    def _fetch_data(self) -> SourceData:
        raise NotImplementedError

    @staticmethod
    def clean_text(text: str) -> str:
        return text.replace("\u202f", " ").replace("\r\n", "\n")


class NotificationGoogleSheetImporter(AbstractNotificationImporter):
    """
    Imports NotificationTemplates from Google Sheets.
    Original source:
    https://github.com/City-of-Helsinki/kukkuu/blob/master/importers/notification_importer.py
    1. Some function and variable renaming done.
    2. Splitted code from original importer to create an abstract class
    """

    TIMEOUT = 5
    HEADER_FIELD_AND_LANGUAGE_SEPARATOR = "|"

    def __init__(self, sheet_id: str = None) -> None:
        self.sheet_id = (
            sheet_id or settings.NOTIFICATIONS_SHEET_ID
            if hasattr(settings, "NOTIFICATIONS_SHEET_ID")
            else None
        )
        if not self.sheet_id:
            raise NotificationImporterException("Sheet ID not set.")

        self.source_data: SourceData = self._fetch_data()

    @property
    def url(self) -> str:
        return (
            f"https://docs.google.com/spreadsheets/d/{self.sheet_id}/export?format=csv"
        )

    def _fetch_data(self) -> SourceData:
        csv_data = self.__fetch_csv_data()
        return self.__get_source_data_from_csv_file(io.StringIO(csv_data))

    def __get_field_and_language_from_header(self, header: str) -> Tuple[str, str]:
        field, language = header.split(self.HEADER_FIELD_AND_LANGUAGE_SEPARATOR)
        return field.lower().strip(), language.lower().strip()

    def __fetch_csv_data(self) -> str:
        try:
            response = requests.get(self.url, timeout=self.TIMEOUT)
            response.raise_for_status()
        except RequestException as e:
            raise NotificationImporterException(
                f"Error fetching data from the spreadsheet: {e}"
            ) from e
        return response.content.decode("utf-8")

    def __get_source_data_from_csv_file(self, csv_file: io.StringIO) -> SourceData:
        reader = csv.DictReader(csv_file)
        source_data = {}

        for row in reader:
            notification_data: DefaultDict[str, Dict[str, str]] = defaultdict(dict)

            row_items = iter(row.items())
            # the first column contains notification type
            notification_type = next(row_items)[1].lower()

            for header, content in row_items:
                field, language = self.__get_field_and_language_from_header(header)
                notification_data[language][field] = self.clean_text(content)

            source_data[notification_type] = notification_data

        return source_data


class NotificationFileImporter(AbstractNotificationImporter):
    """
    Import notification templates from template files.
    The templates files should be stored in notification_importers app in
    notification_importers/templates/sms and notification_importers/templates/email
    folders. There is also a naming convention used there.
    The file name must be given in this pattern [notification_type]-[locale].[html|j2].
    """

    def __init__(self) -> None:
        self.template_dirs = ["/templates/email", "/templates/sms"]
        self.files = sorted(
            [
                filename
                for template_dir in self.template_dirs
                for ext in ("*.html", "*.j2")
                for filename in glob.iglob(
                    os.path.dirname(os.path.abspath(__file__))
                    + template_dir
                    + f"**/{ext}",
                    recursive=True,
                )
            ]
        )
        self.source_data: SourceData = self._fetch_data()

    def _fetch_data(self) -> SourceData:
        source_data: SourceData = defaultdict(lambda: defaultdict(dict))
        for template_file_path in self.files:
            with open(template_file_path, "r") as template_file:
                pathname = os.path.splitext(template_file.name)[0]
                filename = pathname.split("/")[-1]
                notification_type, language = filename.split("-")
                field = (
                    "body_text"
                    if self.__is_sms_template(template_file)
                    else "body_html"
                )
                content = template_file.read()
                source_data[notification_type][language][field] = self.clean_text(
                    content
                )
                source_data[notification_type][language][
                    "subject"
                ] = self.__get_subject(content, notification_type)

        return source_data

    def __get_subject(
        self, file_content: io.TextIOWrapper, notification_type: str
    ) -> str:
        """
        Returns a title element content from a HTML file as a notification subject.
        If a title cannot be found, it uses the file name as a subject.
        """
        title_rgx = re.compile(r"\<title\>(\S.*)\<\/title\>")
        subject = (
            title_rgx.search(file_content).group(1)
            if title_rgx.search(file_content)
            else notification_type.__str__().replace("_", " ")
        )
        return subject

    def __is_sms_template(self, template_file: io.TextIOWrapper) -> bool:
        return True if "sms" in template_file.name else False
