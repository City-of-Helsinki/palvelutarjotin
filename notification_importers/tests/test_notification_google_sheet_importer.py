import pytest
import responses
from django.conf import settings
from django_ilmoitin.models import NotificationTemplate
from occurrences.consts import NotificationTemplate as NotificationType

from common.tests.utils import create_notification_template_in_language

from ..notification_importer import NotificationGoogleSheetImporter
from .utils import serialize_notifications

LANGUAGES = ("fi", "sv", "en")

MOCK_CSV = """,SUBJECT | FI,SUBJECT | SV,SUBJECT | EN,BODY_TEXT | FI,BODY_TEXT | SV,BODY_TEXT | EN\r\n
enrolment_approved,enrolment_approved fi updated subject,enrolment_approved sv updated subject,enrolment_approved en updated subject,enrolment_approved fi updated body_text,enrolment_approved sv updated body_text,enrolment_approved en updated body_text\r\n
occurrence_enrolment,occurrence_enrolment fi updated subject,occurrence_enrolment sv updated subject,occurrence_enrolment en updated subject,occurrence_enrolment fi updated body_text,occurrence_enrolment sv updated body_text,occurrence_enrolment en updated body_text"""  # noqa


@pytest.fixture(autouse=True)
def setup(settings, mocked_responses):
    settings.NOTIFICATIONS_SHEET_ID = "mock-sheet-id"
    settings.PARLER_SUPPORTED_LANGUAGE_CODES = LANGUAGES
    mocked_responses.add(
        responses.GET,
        "https://docs.google.com/spreadsheets/d/mock-sheet-id/export?format=csv",
        body=MOCK_CSV,
        status=200,
        content_type="application/json",
    )


@pytest.fixture
def enrolment_approved_notification():
    for language in LANGUAGES:
        create_notification_template_in_language(
            NotificationType.ENROLMENT_APPROVED,
            language,
            subject=f"{NotificationType.ENROLMENT_APPROVED} "
            f"{language} original subject",
            body_text=f"{NotificationType.ENROLMENT_APPROVED} "
            f"{language} original body_text",
        )
    return NotificationTemplate.objects.get(type=NotificationType.ENROLMENT_APPROVED)


@pytest.fixture
def occurrence_enrolment_notification():
    for language in ("fi", "en"):
        create_notification_template_in_language(
            NotificationType.OCCURRENCE_ENROLMENT,
            language,
            subject=f"{NotificationType.OCCURRENCE_ENROLMENT} "
            f"{language} original subject",
            body_text=f"{NotificationType.OCCURRENCE_ENROLMENT} "
            f"{language} original body_text",
        )
    return NotificationTemplate.objects.get(type=NotificationType.OCCURRENCE_ENROLMENT)


@pytest.mark.django_db
def test_create_non_existing_and_update_existing_notifications(
    enrolment_approved_notification, occurrence_enrolment_notification, snapshot
):
    (
        num_of_created,
        num_of_updated,
    ) = (
        NotificationGoogleSheetImporter().create_missing_and_update_existing_notifications()  # noqa: E501
    )
    assert num_of_created == 0
    assert num_of_updated == 2
    snapshot.assert_match(serialize_notifications(NotificationTemplate.objects.all()))


@pytest.mark.django_db
def test_create_non_existing_notifications(enrolment_approved_notification, snapshot):
    NotificationGoogleSheetImporter().create_missing_notifications()
    snapshot.assert_match(serialize_notifications(NotificationTemplate.objects.all()))


@pytest.mark.django_db
def test_update_notifications(enrolment_approved_notification, snapshot):
    NotificationGoogleSheetImporter().update_notifications(
        [enrolment_approved_notification]
    )
    snapshot.assert_match(serialize_notifications(NotificationTemplate.objects.all()))


@pytest.mark.django_db
def test_is_notification_in_sync(enrolment_approved_notification):
    importer = NotificationGoogleSheetImporter()

    assert not importer.is_notification_in_sync(enrolment_approved_notification)

    # set the notification to match the csv
    for language in settings.PARLER_SUPPORTED_LANGUAGE_CODES:
        translation_obj, _ = enrolment_approved_notification.translations.get_or_create(
            language_code=language
        )
        for field in ("subject", "body_text", "body_html"):
            setattr(
                translation_obj,
                field,
                getattr(translation_obj, field).replace("original", "updated"),
            )
        translation_obj.save()

    assert importer.is_notification_in_sync(enrolment_approved_notification)
