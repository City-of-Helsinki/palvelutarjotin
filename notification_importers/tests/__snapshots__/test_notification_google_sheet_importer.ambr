# serializer version: 1
# name: test_create_non_existing_and_update_existing_notifications
  '''
  enrolment_approved|enrolment_approved fi updated subject|enrolment_approved en updated subject|enrolment_approved sv updated subject|enrolment_approved fi updated body_text|enrolment_approved en updated body_text|enrolment_approved sv updated body_text|||
  occurrence_enrolment|occurrence_enrolment fi updated subject|occurrence_enrolment en updated subject|occurrence_enrolment sv updated subject|occurrence_enrolment fi updated body_text|occurrence_enrolment en updated body_text|occurrence_enrolment sv updated body_text|||
  '''
# ---
# name: test_create_non_existing_notifications
  '''
  enrolment_approved|enrolment_approved fi original subject|enrolment_approved en original subject|enrolment_approved sv original subject|enrolment_approved fi original body_text|enrolment_approved en original body_text|enrolment_approved sv original body_text|||
  occurrence_enrolment|occurrence_enrolment fi updated subject|occurrence_enrolment en updated subject|occurrence_enrolment sv updated subject|occurrence_enrolment fi updated body_text|occurrence_enrolment en updated body_text|occurrence_enrolment sv updated body_text|||
  '''
# ---
# name: test_update_notifications
  'enrolment_approved|enrolment_approved fi updated subject|enrolment_approved en updated subject|enrolment_approved sv updated subject|enrolment_approved fi updated body_text|enrolment_approved en updated body_text|enrolment_approved sv updated body_text|||'
# ---
